Schema v2: Read API integration

dprokop opened the pull request:
Builds on top of https://github.com/grafana/grafana/pull/96666, replaces https://github.com/grafana/grafana/pull/97338

_To test enable `useV2DashboardsAPI` feature toggle._
This adds schema v1->v2 and v2->v1 transformers. v2 API will now transform v1 responses to v2 so that we can move forward with integrating v2 schema into dashboard scene. 

Remaining transformations (added as followup tasks in https://github.com/grafana/grafana-org/issues/239):
- [ ] **v1->v2** Remaining variables 
- [x] **v1->v2** Dashboard id
- [ ] **v2->v1** Annotations
- [ ] **v2->v1** Panels
- [ ] **v2->v1** Variables
- [ ] Rows and repeats
- [ ] Library panels
- [ ] `gnetId` and `revision`

dprokop committed:
V2 schema -> scene meta transformation

dprokop committed:
Force dashbaord scene for v2 api

dprokop committed:
WIP render dashboard from v2 schema

dprokop committed:
v2 schema: Make annotations filter optional

dprokop committed:
v2 schema: Remove defaultOptionEnabled from ds variable schema

dprokop committed:
Update response transformer test

dprokop committed:
Transformation improvements

dprokop committed:
Introduce DashboardScenePageStateManagerLike interface

dprokop committed:
Implement dash loader for handling v2 api

dprokop committed:
v2 api: Handle home dashboard

dprokop committed:
Use correct api client in DashboardScenePage

dprokop committed:
Correctly use v2 dashboard scene serializer

dprokop committed:
Remove unnecesary type assertions

dprokop committed:
Handle v2 dashboard not found

axelavargas reviewed (commented):
Nice progress, I checked the architecture and overall looks great! I like the approach of having the `getDashboardScenePageStateManager` decide which implementation should use depending on the feature flag being enabled. 

It would be great to add to this PR description the diagram from @harisrozajac 

![395727294-c31f5bd1-d5a6-41fa-86fa-52e2195aa5e2](https://github.com/user-attachments/assets/e5baa7d7-9f49-4d7e-bd22-8bbf0228ee01)

harisrozajac commented on public/app/features/dashboard-scene/serialization/transformSceneToSaveModelSchemaV2.ts:
```diff
@@ -405,7 +404,7 @@ function getAnnotations(state: DashboardSceneState): AnnotationQueryKind[] {
         enable: Boolean(layer.state.isEnabled),
         hide: Boolean(layer.state.isHidden),
-        filter: layer.state.query.filter ?? defaultAnnotationPanelFilter(),
+        filter: layer.state.query.filter,
```
When updating snapshot for this and when filter is undefined this is persisted in the snapshot as `filter: undefined`, so maybe we should do something like: 
```
 if (layer.state.query.filter !== undefined) {
      result.spec.filter = layer.state.query.filter;
    }
```
Not sure if there's something in place on the backend or the frontend that cleans up these undefined fields and doesn't persist them in db.

harisrozajac reviewed (approved):
Looks great! Snapshots need updating so the test can pass and I just left one comment regarding that, but those are trivial changes so I'm approving this

axelavargas reviewed (approved):
Hey Dom 😊 , I was giving a try locally, and because of the change of panel.id, I cannot load any dashboard when the feature toggle is enabled.

Code wise I like the architecture of the dashboard manager page. I am approving. As soon as you fix the id issue let me know :), I want to see locally how some dashboards will look like (I know there will be missing pieces, but I believe panels and basic properties should work)

![image](https://github.com/user-attachments/assets/741161d9-9d2e-44c8-9a36-f111c73dbd23)

dprokop committed:
Merge remote-tracking branch 'origin/main' into schemav2-api-interfaces

dprokop committed:
Fix type

dprokop commented on public/app/features/dashboard/api/ResponseTransformers.ts:
```diff
@@ -177,3 +236,217 @@ export const ResponseTransformers = {
+            url: l.url || '',
+            targetBlank: l.targetBlank,
+          })) || [],
+        uid: p.id!.toString(), // TODO[schema v2]: handle undefined id?!!?!?
```
@axelavargas  - it's fixed now :)

dprokop committed:
Fix test

dprokop committed:
Some more tests fix

dprokop committed:
snapshot

dprokop committed:
Add dashboard id annotation

dprokop committed:
Nits

dprokop committed:
Merge branch 'main' into schemav2-api-interfaces

dprokop committed:
Nits

dprokop committed:
Rename v2 api

ivanortegaalba commented on public/app/features/dashboard/api/utils.ts:
```diff
@@ -38,10 +40,16 @@ export function isDashboardResource(
+export function isDashboardV2Resource(
+  obj: DashboardDTO | DashboardWithAccessInfo<DashboardDataDTO> | DashboardWithAccessInfo<DashboardV2Spec>
+): obj is DashboardWithAccessInfo<DashboardV2Spec> {
+  return isDashboardResource(obj) && isDashboardV2Spec(obj.spec);
```
💯

ivanortegaalba commented on public/app/features/dashboard/api/utils.ts:
```diff
@@ -1,12 +1,14 @@
 import { DashboardWithAccessInfo } from './types';
 
 export function getDashboardsApiVersion() {
+  const forcingOldDashboardArch = locationService.getSearch().get('scenes') === 'false';
```
It is worth adding a test case in `utils.test.ts` to avoid breaking the fallback mechanism in the future since this is frequently used to debug scenes issues.

dprokop replied:
added aff7a59bdb1c3523e430161e686282c4fafcf893 :)

ivanortegaalba commented on public/app/features/dashboard/api/ResponseTransformers.test.ts:
```diff
@@ -1,19 +1,52 @@
+        fiscalYearStartMonth: 1,
+        weekStart: 'monday',
+        version: 1,
+        links: [],
```
Let's add some links to ensure we transform them properly since we already support the in the transformers

dprokop replied:
added 84e3e124fdd78502e5d069437e30b56b0269deef :)

ivanortegaalba commented on public/app/features/dashboard/api/ResponseTransformers.test.ts:
```diff
@@ -27,67 +60,53 @@ describe('ResponseTransformers', () => {
+      expect(spec.timeSettings.nowDelay).toBe(dashboardV1.timepicker?.nowDelay);
+      expect(spec.timeSettings.fiscalYearStartMonth).toBe(dashboardV1.fiscalYearStartMonth);
+      expect(spec.timeSettings.weekStart).toBe(dashboardV1.weekStart);
       expect(spec.links).toEqual([]); // Assuming transformDashboardLinksToEnums([]) returns []
```
```suggestion
      expect(spec.links).toEqual([]);
```

ivanortegaalba commented on public/app/features/dashboard/api/ResponseTransformers.ts:
```diff
@@ -1,38 +1,68 @@
   const dashboardDefaults = defaultDashboardV2Spec();
-
-  const dashboard = isDashboardResource(dto) ? dto.spec : dto.dashboard;
+  const [elements, layout] = getElementsFromPanels(dashboard.panels || []);
+  const variables = getVariables(dashboard.templating?.list || []);
+  const annotations = getAnnotations(dashboard.annotations?.list || []);
```
[These transformations are not tested](https://github.com/grafana/grafana/pull/97953/files#diff-b7097f373b2660e4b980ad148346780340b0cf82f6f654b7ea194884e359e186R18), worth adding them to the transformed `DashboardDTO` to be transformed to `DashboardSpec`.

ivanortegaalba commented on public/app/features/dashboard/api/ResponseTransformers.test.ts:
```diff
@@ -1,19 +1,52 @@
 import { DashboardWithAccessInfo } from './types';
 
 describe('ResponseTransformers', () => {
   describe('v1 transformation', () => {
```
```suggestion
  describe('v1 -> v2 transformation', () => {
```
Small nit, just to avoid confusion using symbol navigation 😄 

Same here
https://github.com/grafana/grafana/pull/97953/files#diff-b7097f373b2660e4b980ad148346780340b0cf82f6f654b7ea194884e359e186R115

dprokop replied:
added d222c8122fadd2b41dc05617197bf158f38df646

ivanortegaalba commented on public/app/features/dashboard-scene/serialization/transformSaveModelSchemaV2ToScene.ts:
```diff
@@ -120,9 +162,7 @@ export function transformSaveModelSchemaV2ToScene(dto: DashboardWithAccessInfo<D
     isDirty: false,
     links: dashboard.links,
     // TODO: Combine access and metadata to compose the V1 meta object
```
```suggestion
```
If I'm not wrong, we're already doing this :)

ivanortegaalba reviewed (commented):
All of this, tested with`useV2DashboardsAPI` enabled and disabled
🚧  When disabling scenes with query params, it is called back to v0
🚧  Forcing scenes when featureToggles.useV2DashboardsAPI is enabled
🚧  Dashboard not found behaves correctly
🚧  Opening a dashboard with basic info is rendered correctly

I did some basic testing, I'll continue tomorrow, I commented mostly about testing. The code looks great, the type management is managed in a very clever way. Great work, everything is very easy to understand.

My main concern is `public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts`. We have added a big amount of logic for V1 and V2, and this is not tested, conditionally for the ff enabled or disabled. Is this on purpose? Is this a TODO to be tackled in the next PR?

Great work Dominik!

dprokop committed:
Update public/app/features/dashboard-scene/serialization/transformSaveModelSchemaV2ToScene.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>

dprokop committed:
add getDashboardsApiVersion test for forcing scenes through URL

dprokop committed:
add links to ResponseTransformers

dprokop committed:
Update public/app/features/dashboard/api/ResponseTransformers.test.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>

dprokop committed:
Nit rename test case

dprokop committed:
Add tests for DashboardScenePageStateManagerV2

dprokop commented:
> My main concern is public/app/features/dashboard-scene/pages/DashboardScenePageStateManager.ts. We have added a big amount of logic for V1 and V2, and this is not tested, conditionally for the ff enabled or disabled. Is this on purpose? Is this a TODO to be tackled in the next PR?

@ivanortegaalba  - i've added the tests, they are the same tests as for v1. There are two tests there that are skipped which require a separate work - handling home dash redirect and dashboard not found. First is missing necessary model in the schema - we need to figure out how to handle `redirectUri`, the second one needs a different implementation, as the k8s API can't return undefined dashboard, but throws instead. I think this needs to be addressed in a separate PR, this one is big enough already.

dprokop committed:
Update test

dprokop committed:
Merge branch 'main' into schemav2-api-interfaces

dprokop committed:
Merge remote-tracking branch 'origin/main' into schemav2-api-interfaces

dprokop committed:
Typecheck

dprokop committed:
Add console error for debugging

ivanortegaalba commented:
While testing, I tried to open a regular dashboard and got "Not found." I tried with `scenes=false`, and it worked. So I debugged it, and I got a ResponseTransformer error, which is okay!

<img width="1016" alt="Captura de pantalla 2024-12-31 a las 12 44 33" src="https://github.com/user-attachments/assets/cad5b746-e7a3-4fa4-858c-3955eb7ca838" />

It was hard to debug or understand why it was "Not found". Could we just throw the error in this case, and not assume is not found? Or at least we could throw it in the console. We will get some more ResponseTransformer errors because of missing transformations, but it is hard to understand if we don't get anywhere the real runtime error. WDYT?

ivanortegaalba commented:
✅ When disabling scenes with query params, it is called back to v0
✅ Forcing scenes when `featureToggles.useV2DashboardsAPI` is enabled
✅ Dashboard not found behaves correctly
✅ Opening a dashboard with basic info is rendered correctly (Tested with `useV2DashboardsAPI` enabled and disabled): 
    ✅ Panels, grid positions and viz options
    ⚠️ Transformations not supported yet (Rendering - expected)
    ⚠️ Rows and repeats not supported yet (Dashboard not found - expected)
    ⚠️ Rows and repeats not supported yet (Dashboard not found - expected)
    ✅ Supported variables: query + datasource 
    ⚠️ Unsupported variables  (Dashboard not found - expected)
    ❌  Annotations (See next comment - `useV2DashboardsAPI=true`)
    ✅ Shared queries (Use other panel data)
    ✅ Dashboard link
    ⚠️ Automigrate doesn't work `&__feature.autoMigrateOldPanels=true` added to the URL, the angular panels are not migrated when loading the dashboard—something to note in the epic.
    ✅ Time settings loaded properly

I tested all of this with and without FF enabled, to ensure the correct behavior. Everything works as expected, except annotations. Maybe it is something expected, I just wanted to raise it :)

ivanortegaalba reviewed (approved):
Approving this, since everything works as expected, and the missing features are already documented.

Great job @dprokop!

harisrozajac committed:
Merge branch 'main' into schemav2-api-interfaces

harisrozajac committed:
Fix typo
